### ISSUE TITLE
Overflow

### ISSUE THREAD

**Issue by sanxiyn** [m4b/goblin#23]
Do you care about overflow?

For example, you can trigger one by replacing 8 bytes from 0x20 (e_phoff) of ELF-64 binary with 0xff repeated 8 times.

```
$ RUST_BACKTRACE=1 cargo run --example rdr -- elf
thread 'main' panicked at 'attempt to add with overflow', /home/thomas/.cargo/registry/src/github.com-1ecc6299db9ec823/scroll-0.5.0/src/greater.rs:140
stack backtrace:
  10: <[u8] as scroll::greater::TryOffsetWith<Ctx>>::try_offset
             at /home/thomas/.cargo/registry/src/github.com-1ecc6299db9ec823/scroll-0.5.0/src/greater.rs:140
  11: scroll::greater::Gread::gread_with
             at /home/thomas/.cargo/registry/src/github.com-1ecc6299db9ec823/scroll-0.5.0/src/greater.rs:69
  12: goblin::elf::program_header::std::ProgramHeader::parse
             at src/elf/program_header.rs:142
  13: goblin::elf::impure::Elf::parse
             at src/elf/mod.rs:156
  14: goblin::parse
             at src/lib.rs:273
```

**Comment by m4b** [comment]
Hi @sanxiyn thanks for the issue, and also the backtrace, this helped very quickly figure out what caused this ! :)

So I actually experimented with fuzzing goblin a while back; I'd like to explore that route more, and it did catch some bugs.

In this particular case, iiuc, you're setting `e_phoff = 0xfffffffff`, which means that the program headers start at an offset in a binary larger than the maximum representable size in a `u64`.

In the current version, this overflows because it takes the offset, does not perform any "sanity checks", and tries to parse the header by incrementing by the size of the header + the offset (`0xffffffff`), and so overflows.

I wouldn't mind any changes or PR which does some basic checks, but the `goblin` parser is not very strict when it comes to "well formed" binaries (whatever that actually means, since it's not really specified) - and I'd like to keep it this way.

More importantly, I would _definitely_ be interested in getting a more robust and testable fuzzing experience working using something like AFL (which I used before to catch some basic bugs): https://docs.rs/afl/0.1.5/afl/

If you're interested, let me know, and we can work on getting something setup for making the parser more robust, etc. If you want to start, I'd recommend checking out https://github.com/fitzgen/cpp_demangle which has a good example of integrating it into a project
